gcc/rust/parse/rust-parse-impl.h

```diff
@@ -5614,8 +5614,9 @@ Parser<ManagedTokenSource>::parse_trait_impl_item ()
 	case UNSAFE:
 	case EXTERN_TOK:
 	case FN_TOK:
-	  return parse_trait_impl_function_or_method (visibility,
-						      std::move (outer_attrs));
+	  add_error (Error (t->get_locus (),
+			    "trait functions cannot be declared const"));
+	  return nullptr;
```

Comment:
I'm not sure if this error should come from parsing or from a later phase (like TypeChecking maybe?). I believe const functions in traits might be allowed down the line, and right now they are refused because the various effect systems of Rust are not stabilized. Furthermore, having the error in a later phase would make for a nicer error - and wouldn't say "failed to parse item in crate".

In fact, if we look at the Rust reference for a [Trait](https://doc.rust-lang.org/reference/items/traits.html)

we can see that it expects to parse zero or more [AssociatedItem](https://doc.rust-lang.org/reference/items/associated-items.html), which can be a [Function](https://doc.rust-lang.org/reference/items/functions.html), which has `FunctionQualifier`s. 

So *parsing* a const function in a Trait is allowed, but having a const function in a trait is not allowed - to me this sounds more like a typechecking problem

Reply:
Yeah i think either during HIR lowering you can emit an error or type checking. Or add another check the const checker lint